File: data-conformance/tests/rule.rs
/*
 * This file is part of paged (https://paged.media).
 *
 * paged is free software: you may redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License, version 3, as published by
 * the Free Software Foundation, OR under the Paged Media Enterprise License
 * (PMEL), a commercial license available from And The Next GmbH. Full
 * copyright and license information is available in LICENSE.md, distributed
 * with this source code.
 *
 * paged is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the licenses for details.
 *
 *  @copyright  Copyright (c) And The Next GmbH
 *  @license    AGPL-3.0-only OR Paged Media Enterprise License (PMEL)
 */

/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * This file is part of paged (https://paged.media) and is additionally
 * available under the Paged Media Enterprise License (PMEL). Full
 * copyright and license information is available in LICENSE.md which is
 * distributed with this source code.
 *
 *  @copyright  Copyright (c) And The Next GmbH
 *  @license    MPL-2.0 OR Paged Media Enterprise License (PMEL)
 */

//! Data-driven formatting rule conformance (spec §9.5): the rule's `when`
//! condition is evaluated per record (stabilized order); the firing records get
//! the document-style action. The engine decides WHICH content fires WHICH
//! style — the data-driven half; the host applies the named style (the per-cell
//! application is gated on D-13). Styling is always a document-style reference,
//! never a literal (§9.5).

use data_bind::ResolutionEngine;
use data_conformance::{n, record_set, t, today};
use data_core::{
    Binding, BindingId, FieldType, Query, QueryId, ResultShape, ScopeRef, StyleAction,
};

#[test]
fn data_bind_rule_fires_on_matching_records() {
    let mut e = ResolutionEngine::new(today());
    e.add_query(Query {
        id: QueryId::from("q1"),
        sql: String::new(),
        params: vec![],
        shape: ResultShape::RecordStream,
    });
    e.add_binding(
        BindingId::from("r1"),
        Binding::Rule {
            scope: ScopeRef::from("table-region"),
            when: "stock < 5".into(),
            apply: StyleAction::TableStyle {
                name: "low-stock".into(),
            },
        },
    );
    e.set_result(
        QueryId::from("q1"),
        record_set(
            &[("sku", FieldType::Text), ("stock", FieldType::Float)],
            vec![vec![t("a"), t("b"), t("c")], vec![n(2.0), n(10.0), n(0.0)]],
        ),
    );

    let eval = e
        .evaluate_rule(&BindingId::from("r1"), &QueryId::from("q1"))
        .unwrap();
    // Stabilized by all columns → sku a,b,c. stock 2,10,0 → fires where < 5: a, c.
    assert_eq!(eval.fires, vec![0, 2]);
    assert_eq!(eval.total, 3);
    assert_eq!(eval.scope, ScopeRef::from("table-region"));
    assert_eq!(
        eval.apply,
        StyleAction::TableStyle {
            name: "low-stock".into()
        }
    );
}

#[test]
fn data_bind_rule_non_rule_binding_errors() {
    // evaluate_rule on a non-rule binding is a typed error (it is not a rule).
    let mut e = ResolutionEngine::new(today());
    e.add_query(Query {
        id: QueryId::from("q1"),
        sql: String::new(),
        params: vec![],
        shape: ResultShape::SingleRecord,
    });
    e.add_binding(
        BindingId::from("v1"),
        Binding::Variable {
            target: data_core::PlaceholderRef::from("ph"),
            query: QueryId::from("q1"),
            expr: "name".into(),
            missing: data_core::MissingPolicy::Blank,
        },
    );
    e.set_result(
        QueryId::from("q1"),
        record_set(&[("name", FieldType::Text)], vec![vec![t("x")]]),
    );
    assert!(e
        .evaluate_rule(&BindingId::from("v1"), &QueryId::from("q1"))
        .is_err());
}

File: packages/data-host-model/src/rule.ts
/*
 * This file is part of paged (https://paged.media).
 *
 * paged is free software: you may redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License, version 3, as published by
 * the Free Software Foundation, OR under the Paged Media Enterprise License
 * (PMEL), a commercial license available from And The Next GmbH. Full
 * copyright and license information is available in LICENSE.md, distributed
 * with this source code.
 *
 * paged is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the licenses for details.
 *
 *  @copyright  Copyright (c) And The Next GmbH
 *  @license    AGPL-3.0-only OR Paged Media Enterprise License (PMEL)
 */

// The data-driven rule IR (D-13, spec §9.5) — the exact shape the `data-js`
// engine serialises from `evaluate_rule` (`RuleResult`), plus a pure mapper to
// the host-model's `RuleApplication`. The engine decided WHICH records fired
// WHICH style (the data-driven half); the host applies the named DOCUMENT style
// (never a parallel styling system, never a color literal — §9.5). Type-only +
// pure; the mutations are shaped in `fields.ts` (`ruleMutations`).

/** The engine's `evaluate_rule` output (`RuleResult`, serde camelCase). `apply`
 *  is a tagged document-style action (`action` discriminant, camelCase
 *  variants). */
export interface RuleResult {
  scope: string;
  /** Stabilized record indices where the `when` condition fired. */
  fires: number[];
  apply:
    | { action: "characterStyle"; name: string }
    | { action: "paragraphStyle"; name: string }
    | { action: "tableStyle"; name: string };
  total: number;
}

/** The host-model's normalised rule application: the fired record indices + the
 *  document-style action collapsed to `{kind, name}` (paragraph/character/table
 *  → the host apply path). */
export interface RuleApplication {
  scope: string;
  fires: number[];
  total: number;
  apply:
    | { kind: "character"; name: string }
    | { kind: "paragraph"; name: string }
    | { kind: "table"; name: string };
}

/** Map the engine's `RuleResult` to the host-model `RuleApplication`. Pure: the
 *  firing decision is the engine's; this only re-labels the style action for the
 *  host apply path (`fields.ts` `ruleMutations`). */
export function toRuleApplication(result: RuleResult): RuleApplication {
  const apply =
    result.apply.action === "characterStyle"
      ? ({ kind: "character", name: result.apply.name } as const)
      : result.apply.action === "paragraphStyle"
        ? ({ kind: "paragraph", name: result.apply.name } as const)
        : ({ kind: "table", name: result.apply.name } as const);
  return { scope: result.scope, fires: result.fires, total: result.total, apply };
}
